### ISSUE TITLE
Feature request: Watch window

### ISSUE THREAD

**Issue by tbodt** [Pernosco/pernosco#41]
I'm imagining a window like the local variables window, except you can type your own expressions and they get added to the bottom, and of course then click on them and see dataflow. This is very similar to the "print" expressions on execution windows, except they show the values at the current time instead of at the time of an execution.

**Comment by khuey** [comment]
You can type an expression into the search box to evaluate it today, and if you click on it it will put an entry in the notebook view that has the dataflow/etc stuff you expect. Currently that only evaluates at a fixed time though.

**Comment by tbodt** [comment]
Yeah, I just want that but in a window so I don't have to retype the expression in the search box every time I want to see it.

**Comment by rocallahan** [comment]
Can you describe a bit more the context in which you would use this?

**Comment by tbodt** [comment]
To keep tabs on a global variable, or even `$time`, as I'm moving through time (clicking notebook entries, or executions, or stepping in gdb).

**Comment by rocallahan** [comment]
Thanks.

Can you give a more specific example?

I think some kind of watch window probably makes sense, but I'm also wondering whether there's some way we can zero in on the changes you care about so you don't have to "move through time" manually.

**Comment by tbodt** [comment]
In the case of wanting to look at $time, a view oriented by absolute time would be helpful: my program hanged for a few seconds and then resumed, and I would have liked to see a timeline of execution similar to what you'd see in a profiler, and be able to click a point during the hang to see what was happening there. (I can file a new issue for this if you like.) The time when I wanted to look at a global variable was a while ago so I don't remember why exactly I was doing it or what would have helped find the right point in time faster.

**Comment by rocallahan** [comment]
Mmmm thanks, that is an interesting use-case!

**Comment by rocallahan** [comment]
But I think for figuring out hangs, we can definitely do better than having you watch $time and poke around.

**Comment by khuey** [comment]
For regular global variables we could show all values it ever had pretty easily. That's mostly a UI question. What to do about `$time` is a more interesting question.

**Comment by rocallahan** [comment]
> For regular global variables we could show all values it ever had pretty easily. 

That's actually not that easy since for complex values showing every single byte-level change is probably overkill, they need to be grouped/summarized somehow.

**Comment by rocallahan** [comment]
This hang question seems to demand a view where one axis (probably the X axis) represents time. Then you could click anywhere in it to set the current moment to a specific time.

We could make clicking on a thread name anywhere in the UI added the thread as a row to that view. The thread's row could CPU utilization over time. Clicking on the rendering would of course switch focus to that thread at that moment.

We could also have rows representing all threads in a process or all threads in the system.

For thread rows we could have some toggle that turns it into a flame chart. https://developer.mozilla.org/en-US/docs/Tools/Performance/Flame_Chart
The Firefox profiler would be a good source of inspiration in general for profiler-like features.

**Comment by tbodt** [comment]
That's pretty much the exact thing I was imagining :)

**Comment by rocallahan** [comment]
A lot of people have asked for a view with a time axis in the past. I think most of the time that's actually not what people need, but here, I think it is :-).
